### PR TITLE
remove validation GCE-ism

### LINKED ISSUE #114436: validation has a GCE-ism baked in
pkg/apis/core/validation/validation.go

```
func ValidateReadOnlyPersistentDisks(volumes []core.Volume, fldPath *field.Path) field.ErrorList {                                                                             
    allErrs := field.ErrorList{}
    for i := range volumes {
        vol := &volumes[i]
        idxPath := fldPath.Index(i)
        if vol.GCEPersistentDisk != nil {
            if !vol.GCEPersistentDisk.ReadOnly {
                allErrs = append(allErrs, field.Invalid(idxPath.Child("gcePersistentDisk", "readOnly"), false, "must be true for replicated pods > 1; GCE PD can only be mounted on multiple machines if it is read-only"))
            }    
        }    
        // TODO: What to do for AWS?  It doesn't support replicas
    }    
    return allErrs
}
```

We should probably just get rid of this.

### PR DESCRIPTION
**Pull request by mfordjody**
#### What type of PR is this?
/kind cleanup
/kind api-change

#### What this PR does / why we need it:
remove GCE-ism validation

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/114436

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by mfordjody** [comment]
This pull request https://github.com/kubernetes/kubernetes/pull/114631，I made the latest one.

**Comment by mfordjody** [comment]
This pull request source https://github.com/kubernetes/kubernetes/pull/114631

**Comment by mfordjody** [comment]
cc @thockin

File: pkg/apis/apps/validation/validation.go [hunk 1]
```diff
@@ -725,13 +734,18 @@ func ValidateReplicaSetSpec(spec *apps.ReplicaSetSpec, fldPath *field.Path, opts
 	if err != nil {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "invalid label selector"))
 	} else {
-		allErrs = append(allErrs, ValidatePodTemplateSpecForReplicaSet(&spec.Template, selector, spec.Replicas, fldPath.Child("template"), opts)...)
+		// oldSpec is not empty, pass oldSpec.template.
+		var oldTemplates *api.PodTemplateSpec
```
Comment (thockin) [inline pkg/apis/apps/validation/validation.go:738]: same

File: pkg/apis/apps/validation/validation_test.go [hunk 2]
```diff
@@ -2112,9 +2112,10 @@ func TestValidateDaemonSetUpdate(t *testing.T) {
 	}
 
 	type dsUpdateTest struct {
-		old            apps.DaemonSet
-		update         apps.DaemonSet
-		expectedErrNum int
+		old                             apps.DaemonSet
+		update                          apps.DaemonSet
+		expectedErrNum                  int
+		enableSkipReadOnlyValidationGCE bool
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:2118]: Good tests for DaemonSet - shouldn't we have the equivalent for Deployment and ReplicaSet ?

**Review by thockin — COMMENTED** [review]
Almost there!

File: pkg/apis/apps/validation/validation_test.go (hunk 2 above)
Comment (mfordjody) [reply pkg/apis/apps/validation/validation_test.go:2118]: I only see the replicaset.

Comment (thockin) [reply pkg/apis/apps/validation/validation_test.go:2118]: TestValidateDeployment() ?

**Comment by thockin** [comment]
Code freeze is imminent.  I'll approve in the absence of a test for Deployment, but I'm asking you to please follow-up with one ASAP.

/lgtm
/approve
/retest

File: pkg/apis/apps/validation/validation_test.go (hunk 2 above)
Comment (mfordjody) [reply pkg/apis/apps/validation/validation_test.go:2118]: I try it.

File: pkg/apis/apps/validation/validation_test.go [hunk 3]
```diff
@@ -2720,7 +2753,7 @@ func TestValidateDaemonSet(t *testing.T) {
 	}
 }
 
-func validDeployment() *apps.Deployment {
+func validDeployment(expectedErrNum int, enableSkipReadOnlyValidationGCE bool) *apps.Deployment {
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:2756]: I'm confiused - these are not used anywhere?

Comment (mfordjody) [reply pkg/apis/apps/validation/validation_test.go:2756]: I didn't find a deployment similar testing :(

**Comment by thockin** [comment]
Opportunity to write some?

I approved this without it, but now there's this new change that doesn't
make sense. Revert that part?

File: pkg/apis/apps/validation/validation_test.go [hunk 4]
```diff
@@ -3070,6 +3103,209 @@ func TestValidateDeploymentStatusUpdate(t *testing.T) {
...
+				Spec: apps.DeploymentSpec{
+					Replicas: 2,
+					Selector: &metav1.LabelSelector{MatchLabels: validLabels},
+					Template: readWriteVolumePodTemplate.Template,
+					Strategy: apps.DeploymentStrategy{Type: apps.RecreateDeploymentStrategyType},
+				},
+			},
+		},
+		"invalid read-write volume": {
+			enableSkipReadOnlyValidationGCE: false,
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:3221]: You should either always have this line in each case, or ONLY have this line when setting to true.

**Review by thockin — COMMENTED** [review]
Very close!

File: pkg/apis/apps/validation/validation_test.go [hunk 5]
```diff
@@ -3070,6 +3103,209 @@ func TestValidateDeploymentStatusUpdate(t *testing.T) {
...
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: invalidLabels,
+			},
+		},
+	}
+	type depUpdateTest struct {
+		old                             apps.Deployment
+		update                          apps.Deployment
+		expectedErrNum                  int
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:3148]: As fart as I can tell this is not used anywhere?

File: pkg/apis/apps/validation/validation_test.go [hunk 6]
```diff
@@ -3070,6 +3103,209 @@ func TestValidateDeploymentStatusUpdate(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+				Spec: apps.DeploymentSpec{
+					Replicas: 2,
+					Selector: &metav1.LabelSelector{MatchLabels: validLabels},
+					Template: readWriteVolumePodTemplate.Template,
+					Strategy: apps.DeploymentStrategy{Type: apps.RecreateDeploymentStrategyType},
+				},
+			},
+		},
+		"invalid read-write volume": {
```
[the remark this answers is not in the document]
Comment (thockin) [reply pkg/apis/apps/validation/validation_test.go:3220]: I still can't tell - this is identical to the above case, except that it enables the gate.  What makes it fail?

File: pkg/apis/apps/validation/validation_test.go [hunk 7]
```diff
@@ -3352,8 +3588,10 @@ func TestValidateReplicaSetUpdate(t *testing.T) {
 		},
 	}
 	type rcUpdateTest struct {
-		old    apps.ReplicaSet
-		update apps.ReplicaSet
+		old                             apps.ReplicaSet
+		update                          apps.ReplicaSet
+		expectedErrNum                  int
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:3593]: no used?

File: pkg/apis/apps/validation/validation_test.go [hunk 8]
```diff
@@ -3416,6 +3656,25 @@ func TestValidateReplicaSetUpdate(t *testing.T) {
 				},
 			},
 		},
+		"invalid read-write volume": {
+			enableSkipReadOnlyValidationGCE: true,
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:3660]: I don't see why this would fail?

File: pkg/apis/apps/validation/validation_test.go [hunk 9]
```diff
@@ -3466,12 +3692,16 @@ func TestValidateReplicaSetUpdate(t *testing.T) {
 					Template: validPodTemplate.Template,
 				},
 			},
+			expectedErrNum: 1,
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:3695]: You're setting this field but not actually checking it anywhere?

File: pkg/apis/apps/validation/validation_test.go [hunk 10]
```diff
@@ -3070,6 +3103,192 @@ func TestValidateDeploymentStatusUpdate(t *testing.T) {
...
+			update: apps.Deployment{
+				ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+				Spec: apps.DeploymentSpec{
+					Replicas: -1,
+					Selector: &metav1.LabelSelector{MatchLabels: validLabels},
+					Template: validPodTemplate.Template,
+					Strategy: apps.DeploymentStrategy{Type: apps.RecreateDeploymentStrategyType},
+				},
+			},
+			expectedErrNum: 1,
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:3279]: You're setting this field but not actually checking it anywhere?

**Review by thockin — COMMENTED** [review]
I know writing test code is tedious, but so is reviewing it for the Nth time.  Please take a hard look at the tests here and make sure the test loops are actually convincing us that the function does all that is expected and nothing else.

**Comment by mfordjody** [comment]
I used git reset hard to delete the records submitted incorrectly :(

**Comment by mfordjody** [comment]
I'm sorry for the inconvenience and trouble caused to your work.

**Pull request closed without merging**